_posts/2020-11-30-flexible-and-safe-tuples-in-scala3.md

```diff
@@ -0,0 +1,385 @@
+This is enabled by new and powerful features in the Scala 3 compiler.
+
+In this post we will explore the new capabilities of tuples before
+looking under the hood to learn how the improvements in the Scala 3
+compiler enable to implement type safe operations on tuples.
+
+# The basics: what are tuples ?
+
+In the Python programming language, tuples are a simple concept: 
+they are immutable collections of objects. As such, they are opposed
+to lists, which are mutable.
+
+In Scala both `List`s and tuples are immutable, so why do we care
+about tuples?
+
+Scala being a statically typed programming language, the difference between
+list and tuples is in the type. Lists are *homogeneous* collections while
+tuples are *heterogeneous*. In simpler terms, a tuple collects items maintaining
+the type of each element, while a list collects objects retaining a common type
+for all the elements.
+
+This is better explained with an example:
+```scala
+scala> List(1, "2", 3.0, List(4))
+val res0: List[Any] = List(1, 2, 3.0, List(4))
+```
+We that the compiler tries to infer a common supertype for the elements of the list,
+in this case `Any`.
+
+If do the same with tuples, the elements maintain their individual and specific type:
+```scala
+scala> (1, "2", 3.0, List(4))
+val res0: (Int, String, Double, List[Int]) = (1, 2, 3.0, List(4))
+```
+This behavior is desirable in many cases, for example when
+we want a function to return two or more values or when we want
+to use types to decide which transformation to apply to each
+value.
+
+# How do tuples evolve in Scala 3 ?
+
+## Size limit
+
+Probably the most well know limitation of tuples in Scala 2 was the
+restriction to 22 for the number of elements.
+
+```scala
+scala> (1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23)
+    error: tuples may not have more than 22 elements, but 23 given
+```
+
+In Scala 3 the previous tuple is perfectly legal.
+
+## Element accessor
+
+The only way to retrieve an element of a tuple in Scala 2 was to
+use the (1-based) `._i` attribute. For example:
+
+```scala
+("First", "Second")._2 // "Second"
+```
+
+In Scala 3, we can use the `apply` method with a 0-based argument:
+
+```scala
+("First", "Second")(1) // "Second"
+```
+
+As most of indexes are 0 based in Scala, this brings more consistency
+to codebases. It also provides more flexibility. We can, for example,
+*iterate* on any tuple to print each element on a line:
+
+```scala
+val someStuff = (1, "2", 3.0, List(4))
+for (i <- 0 until someStuff.size)
+  println(someStuff(i))
+```
+This was possible in Scala 2 with the `productIterator` although this
+produced a value of type `Iterator[Any]` which means that we had to pattern
+match or eventually cast the type of the elements.
+
+This brings us to the conceptual change that we will explore in the
+next change: tuples become a collection of data that we can manipulate
+and program against.
```

Comment:
I don't really understand this sentence. We could already program with tuples and manipulate them before.

Reply:
I am trying to say that in Scala 2 tuples are static: we cannot add terms, we cannot drop terms, it is hard to iterate over them etc... while in Scala 3 they are more similar to collections

Reply:
Since the birth of Scala, we've been emphasizing to people learning the language that lists and tuples are completely different things used for completely fferent purposes. Now it seems like the story has changed and they're becoming more interchangeable, more overlapping in purpose. _Why, and under what circumstances, is that even good?_,  is the main thing I feel is missing from this blog post.

Reply:
I don’t have this feeling. I have seen multiple blog articles, talks, or libraries about `HList`s (which is what Scala 3 tuples really are), and I think their usage is quite common in type-level programming. But I agree that we should explain that better in the motivation part of this blog article.

Reply:
Is the difference that application authors won't normally need to manipulate tuples in these new ways, but library authors might?

Reply:
Yeah, that's probably more common to library authors.